### PR TITLE
`set_epoch` for validation and prediction data loaders

### PR DESCRIPTION
**Pull request by krshrimali**
## What does this PR do?

Also working on adding a warning to suggest that using `set_epoch` is not suggested during validation/testing.
 

Fixes https://github.com/PyTorchLightning/pytorch-lightning/issues/10342

### Does your PR introduce any breaking changes? If yes, please list them.

None

## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/loops/epoch/evaluation_epoch_loop.py [hunk 1]
```diff
@@ -150,6 +150,14 @@ def advance(  # type: ignore[override]
             # if fault tolerant is enabled and process has been notified, exit.
             self.trainer._exit_gracefully_on_signal()
 
+    def on_advance_start(self) -> None:  # type: ignore[override]
+        if self.trainer.val_dataloader is not None and callable(
```
Comment (carmocca) [inline pytorch_lightning/loops/epoch/evaluation_epoch_loop.py:154]: You will need to select the dataloader based on `if self.trainer.testing`

File: pytorch_lightning/loops/epoch/prediction_epoch_loop.py [hunk 2]
```diff
@@ -103,6 +103,14 @@ def advance(  # type: ignore[override]
 
         self._predict_step(batch, batch_idx, dataloader_idx)
 
+    def on_advance_start(self) -> None:  # type: ignore[override]
+        if self.trainer.test_dataloader is not None and callable(
```
Comment (carmocca) [inline pytorch_lightning/loops/epoch/prediction_epoch_loop.py:107]: `trainer.predict` uses `trainer.predict_dataloaders`

File: pytorch_lightning/loops/epoch/evaluation_epoch_loop.py [hunk 3]
```diff
@@ -150,6 +150,14 @@ def advance(  # type: ignore[override]
             # if fault tolerant is enabled and process has been notified, exit.
             self.trainer._exit_gracefully_on_signal()
 
+    def on_advance_start(self) -> None:  # type: ignore[override]
```
[the remark this answers is not in the document]
Comment (krshrimali) [reply pytorch_lightning/loops/epoch/evaluation_epoch_loop.py]: You are right, fixed now :)

File: pytorch_lightning/loops/epoch/evaluation_epoch_loop.py (hunk 1 above)
Comment (krshrimali) [reply pytorch_lightning/loops/epoch/evaluation_epoch_loop.py]: Used `self.dataloaders` instead.

File: pytorch_lightning/loops/epoch/prediction_epoch_loop.py (hunk 2 above)
Comment (krshrimali) [reply pytorch_lightning/loops/epoch/prediction_epoch_loop.py]: Done! Thanks

File: pytorch_lightning/loops/dataloader/prediction_loop.py [hunk 4]
```diff
@@ -96,6 +96,18 @@ def advance(self, *args: Any, **kwargs: Any) -> None:
         self.predictions.append(dl_predictions)
         self.epoch_batch_indices.append(dl_batch_indices)
 
+    def on_advance_start(self) -> None:  # type: ignore[override]
+        dataloader = self.dataloaders[self.current_dataloader_idx]
+        if (
+            dataloader is not None
+            and getattr(dataloader, "sampler", None)
+            and callable(getattr(dataloader, "set_epoch", None))
```
Comment (awaelchli) [inline pytorch_lightning/loops/dataloader/prediction_loop.py:104]:
```suggestion
            and callable(getattr(dataloader.sampler, "set_epoch", None))
```

File: pytorch_lightning/loops/dataloader/evaluation_loop.py [hunk 5]
```diff
@@ -161,6 +161,18 @@ def advance(self, *args: Any, **kwargs: Any) -> None:
             # indicate the loop has run
             self._has_run = True
 
+    def on_advance_start(self) -> None:  # type: ignore[override]
```
Comment (awaelchli) [inline pytorch_lightning/loops/dataloader/evaluation_loop.py:164]: to avoid the 
`# type: ignore[override]`
you can simply add the args kwargs

Comment (krshrimali) [reply pytorch_lightning/loops/dataloader/evaluation_loop.py]: Done! Thanks for the suggestion, @awaelchli.

Comment (krshrimali) [reply pytorch_lightning/loops/dataloader/evaluation_loop.py]: Should be fixed now :)

**Review by ananthsub — APPROVED** [review]

**Comment by krshrimali** [comment]
> @krshrimali are you planning to add a test? Simplest would probably be to do it with a Mock to check the calls.

Thank you so much, @awaelchli for helping me with the tests (offline) ❤️ I have updated this PR, and would really appreciate it if you can take a look!

File: tests/loops/test_evaluation_loop.py [hunk 6]
```diff
@@ -42,6 +44,28 @@ def test_on_evaluation_epoch_end(eval_epoch_end_mock, tmpdir):
...
+    )
+
+    trainer.fit(model, dataloader)
+    # One for each epoch!
+    assert dataloader.sampler.set_epoch.call_count == 2
+
+    predict_dataloader = torch.utils.data.dataloader.DataLoader(dataset, sampler=sampler)
+    trainer.predict(model, predict_dataloader)
+    # One for each epoch!
+    assert predict_dataloader.sampler.set_epoch.call_count == 2
```
Comment (awaelchli) [inline tests/loops/test_evaluation_loop.py:66]: This only returns 2 because of the previous value that the mock is holding from fit(). 
If you `sampler.reset_mock()` then it returns 0. 
This is because the trainer replaces the sampler no matter what the user sets there (shuffle=True or sampler=). 

I looked at the linked issue again and there the user wants to shuffle the data for validation and test. This is currently not possible and setting the epoch for the distributed sampler won't matter, because shuffling is turned off. 

The issue is actually directly related to #11856. So without directly resolving a decision for #11856, setting the epoch here won't matter. Sorry that I didn't spot this relation earlier. I should have noticed.

Comment (rohitgr7) [reply tests/loops/test_evaluation_loop.py:66]: >  This is currently not possible and setting the epoch for the distributed sampler won't matter, because shuffling is turned off.

for val/test/predict, shuffle will depend upon the dataloader passed and will be used will initializing the `DistributedSampler`.

**Review by rohitgr7 — APPROVED** [review]

File: pytorch_lightning/trainer/connectors/data_connector.py [hunk 7]
```diff
@@ -500,8 +500,8 @@ def replace_sampler(dataloader: DataLoader) -> DataLoader:
     def _check_eval_shuffling(dataloader, mode):
         if _is_dataloader_shuffled(dataloader):
             rank_zero_warn(
-                f"Your `{mode.dataloader_prefix}_dataloader` has `shuffle=True`,"
-                " it is strongly recommended that you turn this off for val/test/predict dataloaders.",
+                f"Your `{mode.dataloader_prefix}_dataloader`'s sampler has shuffling enabled,"
+                " it is strongly recommended that you turn shuffling off for val/test/predict dataloaders.",
```
Comment (krshrimali) [inline pytorch_lightning/trainer/connectors/data_connector.py:504]: This partially addresses the issue https://github.com/PyTorchLightning/pytorch-lightning/issues/11856.

File: tests/loops/test_evaluation_loop.py [hunk 8]
```diff
@@ -42,6 +44,37 @@ def test_on_evaluation_epoch_end(eval_epoch_end_mock, tmpdir):
...
+    model = BoringModel()
+
+    train_dataloader = _get_dataloader()
+    val_dataloader = _get_dataloader()
+
+    trainer = Trainer(
+        default_root_dir=tmpdir, limit_train_batches=2, limit_val_batches=2, max_epochs=2, enable_model_summary=False
+    )
+
+    trainer.fit(model, train_dataloaders=train_dataloader, val_dataloaders=val_dataloader)
```
Comment (krshrimali) [inline tests/loops/test_evaluation_loop.py:67]: We need to pass the data loaders (for training and validation) separately else the model's data loader is considered for evaluation. 

https://github.com/PyTorchLightning/pytorch-lightning/blob/ae4bd0354548d1234dbbfb2acce95d6323ec6b91/pytorch_lightning/trainer/connectors/data_connector.py#L187-L189

This fixes the error, and `set_epoch` now works as expected. :)

**Review by Borda — APPROVED** [review]
lgtm

**Review by carmocca — APPROVED** [review]

File: tests/loops/test_evaluation_loop.py [hunk 9]
```diff
@@ -42,6 +44,37 @@ def test_on_evaluation_epoch_end(eval_epoch_end_mock, tmpdir):
...
 
+def test_set_epoch_called_eval_predict(tmpdir):
+    """Tests that set_epoch (if the sampler has one) is called on the DataLoader during evaluation and
+    prediction."""
+
+    def _get_dataloader():
+        dataset = RandomDataset(32, 64)
+        sampler = RandomSampler(dataset)
+        sampler.set_epoch = Mock()
+        sampler.set_epoch.reset_mock()
```
[the remark this answers is not in the document]
Comment (krshrimali) [reply tests/loops/test_evaluation_loop.py]: Ah, right! good catch, thanks @carmocca!!
